### ISSUE TITLE
NRE in System.Windows.Controls.PopupControlService.RaiseToolTipClosingEvent(bool reset)

### ISSUE THREAD

**Issue by clairernovotny** [dotnet/wpf#381]
Windows 10 17134

WPF version: `4.8.27325.3 @BuiltBy: dlab14-DDVSOWINAGE058 @Branch: master @Commit: 24b4ae9332545f7677a8ce2f8916cc2284e31b75`

This resulted in an app crash as it was unhandled

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Windows.Controls.PopupControlService.RaiseToolTipClosingEvent(bool reset)
   at System.Windows.Controls.PopupControlService.InspectElementForToolTip(DependencyObject o, ToolTipTrigger triggerAction)
   at System.Windows.Controls.PopupControlService.OnFocusChanged(object sender, KeyboardFocusChangedEventArgs e)
   at System.Windows.Input.KeyboardNavigation+<>c__DisplayClass22_0.<NotifyFocusChanged>b__0(object item)
   at System.Windows.Input.KeyboardNavigation+WeakReferenceList.Process(Func<T, TResult> action)
   at System.Windows.Input.KeyboardNavigation.NotifyFocusChanged(object sender, KeyboardFocusChangedEventArgs e)
   at System.Windows.FrameworkElement.OnGotKeyboardFocus(object sender, KeyboardFocusChangedEventArgs e)
   at System.Windows.Input.KeyboardFocusChangedEventArgs.InvokeEventHandler(Delegate genericHandler, object genericTarget)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(object source, RoutedEventArgs args, bool reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseTrustedEvent(RoutedEventArgs args)
   at System.Windows.Input.InputManager.ProcessStagingArea()
   at System.Windows.Input.InputManager.ProcessInput(InputEventArgs input)
   at System.Windows.Input.KeyboardDevice.ChangeFocus(DependencyObject focus, int timestamp)
   at System.Windows.Input.KeyboardDevice.TryChangeFocus(DependencyObject newFocus, IKeyboardInputProvider keyboardInputProvider, bool askOld, bool askNew, bool forceToNullIfFailed)
   at System.Windows.Input.KeyboardDevice.Focus(DependencyObject focus, bool askOld, bool askNew, bool forceToNullIfFailed)
   at System.Windows.Input.KeyboardDevice.Focus(IInputElement element)
   at System.Windows.Input.Keyboard.Focus(IInputElement element)
   at System.Windows.Interop.HwndKeyboardInputProvider.OnSetFocus(IntPtr hwnd)
   at System.Windows.Interop.HwndKeyboardInputProvider.FilterMessage(IntPtr hwnd, WindowMessage message, IntPtr wParam, IntPtr lParam, ref bool handled)
   at System.Windows.Interop.HwndSource.InputFilterMessage(IntPtr hwnd, int msg, IntPtr wParam, IntPtr lParam, ref bool handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, int msg, IntPtr wParam, IntPtr lParam, ref bool handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, object args, int numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, Delegate callback, object args, int numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, object args, int numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, int msg, IntPtr wParam, IntPtr lParam)
```

**Comment by stevenbrix** [comment]
Thanks for filing this issue @onovotny. Do you have a repro that we can use to investigate?

**Comment by onovotny** [comment]
I don't have a specific repro, this came from telemetry.

**Comment by stevenbrix** [comment]
This is an interesting bug, from inspecting the code i'm not quite sure where an NRE would come from. Moving this to future and hopefully we can get more information.

**Comment by YoshihiroIto** [comment]
I also have this problem in my project.
And I made a minimal reproduction project.


https://github.com/YoshihiroIto/NRE_RaiseToolTipClosingEvent



procedure
1. Click “Show Dialog” button
=> "Dialog" window opens

2. Click “Test” button
=> Browser opens

3. Click “MainWindow”
=> Window is in front

4. Click “MainWindow” again
=> NRE

**Comment by YoshihiroIto** [comment]
https://github.com/YoshihiroIto/NRE_RaiseToolTipClosingEvent/blob/693ff9865b7a8b2e3e65a11f6afe8a3b45df7da2/NRE_RaiseToolTipClosingEvent/MainWindow.xaml.cs#L16

Without this line, no exception is raised.
Is there a problem with the handling of the owner window?

I cannot determine if this is a user program issue or a WPF issue.
What do you think?

**Comment by weltkante** [comment]
I don't see any user mistake with that line, its probably a bug.

**Comment by YoshihiroIto** [comment]
https://github.com/YoshihiroIto/NRE_RaiseToolTipClosingEvent/blob/693ff9865b7a8b2e3e65a11f6afe8a3b45df7da2/NRE_RaiseToolTipClosingEvent/App.xaml#L7

This line is also a necessary condition to cause NRE too.

**Comment by singhashish-wpf** [comment]
This issue has been addressed in the later releases (.NET Core 6+)

**Comment by jonathbil** [comment]
How a bug in Framework 4.8 not fix? Not every project can be ported to .Net Core 6+. Also based on MSFT commitment are they supposed to support (I understand fix) Framework 4.8 for many years ahead?

**Comment by lindexi** [comment]
@jonathbil I think it fix in .NET Framework 4.8.

**Comment by singhashish-wpf** [comment]
> How a bug in Framework 4.8 not fix? Not every project can be ported to .Net Core 6+. Also based on MSFT commitment are they supposed to support (I understand fix) Framework 4.8 for many years ahead?

Yes, the fix will be in 4.8.1
